paddle/fluid/platform/collective_helper.cc

```diff
@@ -159,7 +158,159 @@ void NCCLCommContext::ReleaseNCCLComms() {
   }
 }
 
-}  // namespace platform
-}  // namespace paddle
+#endif
+
+#if defined(PADDLE_WITH_XPU_BKCL)
+
+class BKCLCommImpl : public BKCLComm {
+ public:
+  void set_ring_id(int ring_id) { ring_id_ = ring_id; }
+  int ring_id() const override { return ring_id_; }
+
+  void set_nranks(int nranks) { nranks_ = nranks; }
+  int nranks() const override { return nranks_; }
+
+  void set_rank(int rank) { rank_ = rank; }
+  int rank() const override { return rank_; }
+
+  int device_id() const override {
+    return BOOST_GET_CONST(XPUPlace, dev_ctx_->GetPlace()).device;
+  }
+
+  void set_comm(BKCLContext_t comm) { comm_ = comm; }
+  BKCLContext_t comm() const override { return comm_; }
+
+  XPUStream stream() const override {
+    return dev_ctx_->x_context()->xpu_stream;
+  }
+
+  void set_dev_ctx(std::unique_ptr<XPUDeviceContext>&& dev_ctx) {
+    dev_ctx_ = std::move(dev_ctx);
+  }
+  XPUDeviceContext* dev_context() const override { return dev_ctx_.get(); }
+
+ private:
+  int ring_id_;
+  int nranks_;
+  int rank_;
+  BKCLContext_t comm_;
+  std::unique_ptr<XPUDeviceContext> dev_ctx_;
+};
+
+BKCLComm* BKCLCommContext::CreateBKCLComm(BKCLUniqueId* bkcl_id, int nranks,
+                                          int rank, int dev_id, int ring_id) {
+  PADDLE_ENFORCE_NOT_NULL(bkcl_id,
+                          platform::errors::InvalidArgument(
+                              "The bkcl unique id should not be null."));
+  PADDLE_ENFORCE_GT(
+      nranks, 1,
+      platform::errors::InvalidArgument(
+          "Expected nranks > 1. But received nranks is %d.", nranks));
+  PADDLE_ENFORCE_GE(rank, 0,
+                    platform::errors::InvalidArgument(
+                        "Expected rank >= 0. But received rank is %d.", rank));
+  PADDLE_ENFORCE_LT(
+      rank, nranks,
+      platform::errors::InvalidArgument(
+          "Expected rank < nranks. But received rank is %d, nranks is %d.",
+          rank, nranks));
+  PADDLE_ENFORCE_GE(
+      dev_id, 0,
+      platform::errors::InvalidArgument(
+          "Expected dev_id >= 0. But received dev_id is %d.", dev_id));
+
+  BKCLContext_t comm = nullptr;
+  PADDLE_ENFORCE_EQ(
+      xpu_set_device(dev_id), XPU_SUCCESS,
+      platform::errors::PreconditionNotMet("xpu_set_device failed"));
+
+  PADDLE_ENFORCE_EQ(
+      bkcl_init_rank(&comm, rank, nranks, bkcl_id), BKCL_SUCCESS,
+      platform::errors::PreconditionNotMet("bkcl_init_rank failed"));
+
+  auto* comm_wrapper = AssignBKCLComm(comm, nranks, rank, dev_id, ring_id);
+
+  VLOG(1) << "bkcl communicator of rank " << rank << " in ring " << ring_id
+          << " has been created on device " << dev_id;
+
+  std::call_once(once_flag_, []() {
+    std::atexit([]() { BKCLCommContext::Instance().ReleaseBKCLComms(); });
+  });
+
+  return comm_wrapper;
+}
+
+/*
```

Comment:
Remove these useless lines? Or will Kunlun support this later?

Reply:
Done